man/smplayer.1
.TH smplayer 1 "January 2012" "The SMPlayer Project" "SMPlayer"
.SH NAME
SMPlayer \- The best GUI frontend for MPlayer
.SH SYNOPSIS
.B smplayer
.I [\fI-minigui\fR] [\fI-mpcgui\fR] [\fI-defaultgui\fR] [\fI-config-path <directory>\fR]
[\fI-send-action <action_name>\fR] [\fI-actions <action_list>\fR] [\fI[\-no]\-close\-at\-end\fR] 
[\fI[\-no]\-fullscreen\fR] [\fI\-sub <subtitle_file>\fR] [\fI-pos <x> <y>\fR] [\fI-size <width> <height>\fR] 
[\fI\-add\-to\-playlist\fR] [\fI-disable-server] 
[\fI\-help|\-\-help|\-h|\-?\fR] [\fI<media1>\fR] [\fI<media2>\fR]...
.SH DESCRIPTION
.TP
SMPlayer is a GUI media player based on Qt, using \fBmplayer\fR(1) as its backend.
.SH OPTIONS
.TP
.B \-minigui
Open the mini GUI instead of the default one.
.TP
.B \-mpcgui
Open the mpc GUI instead of the default one.
.TP
.B \-defaultgui
Open the default gui.
.TP
.B \-config\-path <directory>
Specify the directory for the configuration files (smplayer.ini, smplayer_files.ini...).
.TP
.B \-send\-action <action_name>
Try to make a connection to another running instance and send to it the
specified action. 
.IP
\fIExample\fR: smplayer \-send\-action pause
.IP
The rest of options (if any) will be ignored and the application will exit. It will return 0
on success or \-1 on failure.
.TP
.B \-actions <action_list>
<action_list> is a list of actions separated by spaces. The
actions will be executed just after loading the file (if
any) in the same order as input. For switchable actions
you can pass true or false as parameter. 
.IP
\fIExample\fR: smplayer \-actions.br "fullscreen compact true"
.IP
Quotes are necessary in case you pass more than one action.
.TP
.B \-close\-at\-end
Close the main window when the file/playlist
finishes.
.TP
.B \-no\-close\-at\-end
Do not close the main window when the file/playlist
finishes.
.TP
.B \-fullscreen
Play the video in fullscreen mode.
.TP
.B \-no\-fullscreen
Play the video in windowed mode.
.TP
.B \-sub
Specify the subtitle file to be loaded (for the first video only).
.TP
.B \-pos <x> <y>
Specify the coordinates where the main window will be displayed.
.TP
.B \-size <width> <height>
Specify the size of the main window.
.TP
.B \-h, \-help, \--help, \-?
Show the help message and then exit.
.TP
.B -add\-to\-playlist
If there is another instance running, the media will be
added to that instance's playlist. If there is no other
instance, this option will be ignored and the files will be
opened in a new instance.
.TP
.B -disable-server
Disables the server used to communicate with other instances. 
This also disables the possibility to use a single instance.
.TP
.B <media1> <media2> ...
Any kind of file that SMPlayer can open. It can be a local file,
a DVD (e.g. dvd://1), an Internet stream
(e.g. mms://....) or a local playlist in format m3u or pls.
.SH "SEE ALSO"
\fBmplayer\fR(1)
.SH AUTHOR
The author of SMPlayer is Ricardo Villalba <ricardo@smplayer.info>.

This manual page was written by Matvey Kozhev <sikon@lucidfox.org> for the
Ubuntu project (but may be used by others).
Has been updated by greengreat <gmeildeno@gmail.com> for the SMPlayer project (July 2009).

Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU General Public License, Version 2 or any later
version published by the Free Software Foundation. On Debian systems, the
complete text of the GNU General Public License can be found in
\fB/usr/share/common-licenses/GPL\fR.
				   
.SH BUGS
Please submit bugs to \fBhttp://sourceforge.net/projects/smplayer\fR.


Changelog
Version 21.1:

(2020-11-28)
 * Try to prevent a crash when using the actions play next/prev very fast.

(2020-11-27)
 * (Playlist) Fix position number after adding files.
 * Hide the shuffle column by default.

(2020-11-18)
 * Pass the option gpu-context when using wayland.
 
(2020-11-09)
 * Add mpsub to the extension list.
 
(2020-10-26)
 * (Find subtitles) Add filters for Spanish and Portuguese languages.

(2020-10-22)
 * Add cache in seconds and cache speed to the OSD info.

(2020-10-21)
 * Add option to display the remaining time in the status bar.

(2020-10-07)
 * Add the packages youtube-dl and python-is-python3 to the "recommends"
   section in the deb package.

(2020-09-14)
 * Try to prevent the video outside the GUI when using Wayland.
 * The shuffle button in the playlist window can be turned on/off.


Version 20.6:

(2020-05-19)
 * New option in the help menu to help install youtube-dl.

(2020-05-18)
 * The internal youtube code has been removed and now smplayer just
   uses youtube-dl to get the video urls.

(2020-04-18)
 * The shuffle option in the playlist does shuffle the items.

(2020-04-15)
 * Set a white background in the qr code label.


Version 20.4.2:

(2020-04-12)
 * Fix youtube titles.

(2020-04-10)
 * Fix youtube.
 * Don't use Qt script.


Version 20.4.0:

(2020-04-05)
 * Possibility to enter the opensubtitles credentials.

(2020-03-31)
 * Use IPC with mpv.

(2020-02-07)
 * Fix for mpv 0.32.

(2019-12-15)
 * Possibility to search for shortcuts in the shortcut editor.
 * Bug fix: the "previous subtitle" action didn't have a name.


Version 19.10.2:

(2019-11-05)
 * (Linux) Try to prevent the player window to be larger than the desktop
   after a resize.

(2019-11-04)
 * Fix compiling the web server on 32 bit with gcc 9.2.0.


Version 19.10:

(2019-10-26)
 * Fix the control problem with mpv 0.30.
 * Set minimum icon size for toolbars to 8 pixels.

(2019-09-19)
 * Allow negative margins for subtitles (mplayer only).
 * (Chromecast) Set overwrite_vtt to true and sub_position to 95.

(2019-09-13)
 * Fix youtube.

(2019-07-08)
 * New audio extensions: mp2, mpc, wv, shn.


Version 19.5:

(2019-05-10)
 * Fix for YouTube.

(2019-03-18)
 * Try to prevent to play invalid YouTube streams.

(2019-03-11)
 * Fix YouTube live streams (it seems to work only with mpv).

(2019-02-25)
 * Add actions for previous subtitle track, previous audio track
   and previous video track.

(2019-02-23)
 * Add command line option -start.


Version 19.1:

(2019-01-27)
 * Fix pause in mpris2.

(2019-01-26)
 * Possibility to play YouTube videos at 60 fps.

(2019-01-08)
 * Don't load the external audio file if it doesn't exist.

(2018-11-22)
 * Possibility to search for subtitles by title name.


Version 18.10:

(2018-10-09)
 * Fixed the YouTube age restricted videos.

(2018-10-07)
 * Experimental code which can decrypt the YouTube signatures without
   running javascript code.

(2018-10-01)
 * Add the option "Use adaptive streams" in Preferences -> Network.

(2018-09-30)
 * The internal code for YouTube now supports DASH streams, which provide
   videos with better resolutions.


Version 18.9:

(2018-09-08)
  * Fix Youtube.

(2018-09-05)
 * Display an icon in the system tray when the SMPlayer web server is running.

(2018-09-03)
 * Add option Play -> Cast to -> Smartphone/tablet. It displays a QR code
   that you can scan to play the video on your device.

(2018-08-31)
 * Add truehd to the audio-spdif option.

(2018-08-29)
 * (Chromecast) Change the default value for the position of the subtitles
   to -1 (default position).


Version 18.6:

(2018-06-18)
 * List all video tracks in the View Information dialog.

(2018-06-13)
 * Fix the "busy" cursor when using the functions play previous/next
   on Windows 10.

(2018-06-05)
 * Add an option in the playlist preferences to select if the playlist is
   dockable inside the main window (old behaviour) or just a normal
   window (new).

(2018-05-27)
 * Update the mongoose code which is used by the web server for the chromecast
   feature. The old one failed to compiled with newer versions of gcc and
   it seems it had several vulnerabilities.
 * Add option use_player_shortcuts_in_playlist for smplayer.ini.


Version 18.5:

(2018-05-20)
 * Add "opus" to the list of audio extensions and "part" to the list of
   video extensions.

(2018-05-19)
 * The cursor keys work now in the playlist (as long as the playlist
   is in a window and not docked).

(2018-05-17)
 * Fix compiling with Qt 5.11.

(2018-05-15)
 * Now it's possible to select which multimedia keys should be used as
   global shortcuts.

(2018-05-01)
 * (Windows) It seems on some systems the multimedia key events are sent
   twice by Qt, making some keys (like volume mute or play/pause) unusable.
   Added code to try to detect this and ignore the second event.

(2018-04-30)
 * (Windows) Use the Windows GetCommandLine function to get the command line
   arguments because it seems the Qt arguments function is not reliable to
   get unicode characters.


Version 18.4:

(2018-04-22)
 * Report errors in the status bar if the option to display a dialog for
   errors is disabled.
 * Set the option report_player_crashes to false by default on Windows.

(2018-04-21)

src/smplayer.cpp
/*  smplayer, GUI front-end for mplayer.
    Copyright (C) 2006-2021 Ricardo Villalba <ricardo@smplayer.info>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include "smplayer.h"
#include "global.h"
#include "paths.h"
#include "translator.h"
#include "version.h"
#include "config.h"
#include "clhelp.h"
#include "cleanconfig.h"
#include "myapplication.h"
#include "baseguiplus.h"

#ifdef DEFAULTGUI
#include "defaultgui.h"
#endif

#ifdef MINIGUI
#include "minigui.h"
#endif

#ifdef MPCGUI
#include "mpcgui.h"
#endif

#ifdef SKINS
#include "skingui.h"
#endif

#include <QDir>
#include <QUrl>
#include <QTime>
#include <stdio.h>

#ifdef Q_OS_WIN
#if USE_ASSOCIATIONS
#include "extensions.h"
#include "winfileassoc.h"	//required for Uninstall
#endif
#endif

#ifdef FONTCACHE_DIALOG
#include "fontcache.h"
#include "version.h"
#endif

using namespace Global;

BaseGui * SMPlayer::main_window = 0;

SMPlayer::SMPlayer(const QString & config_path, QObject * parent )
	: QObject(parent)
	, initial_second(0)
{
#ifdef LOG_SMPLAYER
	#if QT_VERSION >= 0x050000
	qInstallMessageHandler( SMPlayer::myMessageOutput );
	#else
	qInstallMsgHandler( SMPlayer::myMessageOutput );
	#endif
	allow_to_send_log_to_gui = true;
#endif

	gui_to_use = "DefaultGUI";

	close_at_end = -1; // Not set
	start_in_fullscreen = -1; // Not set

	move_gui = false;
	resize_gui = false;

	Paths::setAppPath( qApp->applicationDirPath() );

#ifndef PORTABLE_APP
	if (config_path.isEmpty()) createConfigDirectory();
#endif
	global_init(config_path);

	// Application translations
	translator->load( pref->language );
	showInfo();

#ifdef FONTS_HACK
	createFontFile();
#endif
}

SMPlayer::~SMPlayer() {
	if (main_window != 0) {
		deleteGUI();
	}
	global_end();

#ifdef LOG_SMPLAYER
	if (output_log.isOpen()) output_log.close();
#endif
}

BaseGui * SMPlayer::gui() {
	if (main_window == 0) {
		// Changes to app path, so smplayer can find a relative mplayer path
		QDir::setCurrent(Paths::appPath());
		qDebug("SMPlayer::gui: changed working directory to app path");
		qDebug("SMPlayer::gui: current directory: %s", QDir::currentPath().toUtf8().data());

#ifdef SKINS
		if (gui_to_use == "SkinGUI") {
			QString theme = pref->iconset;
			if (theme.isEmpty()) theme = "Gonzo";
			QString user_theme_dir = Paths::configPath() + "/themes/" + theme;
			QString theme_dir = Paths::themesPath() + "/" + theme;
			qDebug("SMPlayer::gui: user_theme_dir: %s", user_theme_dir.toUtf8().constData());
			qDebug("SMPlayer::gui: theme_dir: %s", theme_dir.toUtf8().constData());
			if ((QDir(theme_dir).exists()) || (QDir(user_theme_dir).exists())) {
				if (pref->iconset.isEmpty()) pref->iconset = theme;
			} else {
				qDebug("SMPlayer::gui: skin folder doesn't exist. Falling back to default gui.");
				gui_to_use = "DefaultGUI";
				pref->iconset = "";
				pref->gui = gui_to_use;
			}
		}
#endif

		main_window = createGUI(gui_to_use);

		if (move_gui) {
			qDebug("SMPlayer::gui: moving main window to %d %d", gui_position.x(), gui_position.y());
			main_window->move(gui_position);
		}
		if (resize_gui) {
			qDebug("SMPlayer::gui: resizing main window to %dx%d", gui_size.width(), gui_size.height());
			main_window->resize(gui_size);
		}
	}

	return main_window;
}

BaseGui * SMPlayer::createGUI(QString gui_name) {
	BaseGui * gui = 0;

#ifdef SKINS
	if (gui_name.toLower() == "skingui")
		gui = new SkinGui(0);
	else
#endif
#ifdef MINIGUI
	if (gui_name.toLower() == "minigui") 
		gui = new MiniGui(0);
	else
#endif
#ifdef MPCGUI
	if (gui_name.toLower() == "mpcgui")
		gui = new MpcGui(0);
	else
#endif
#ifdef DEFAULTGUI
	if (gui_name.toLower() == "defaultgui")
		gui = new DefaultGui(0);
	else
#endif
	{
		// No GUI
		qWarning() << "SMPlayer::createGUI: there's no GUI available!";
		gui = new BaseGuiPlus(0);
	}

	gui->setForceCloseOnFinish(close_at_end);
	gui->setForceStartInFullscreen(start_in_fullscreen);
	connect(gui, SIGNAL(quitSolicited()), qApp, SLOT(quit()));

#ifdef GUI_CHANGE_ON_RUNTIME
	connect(gui, SIGNAL(guiChanged(QString)), this, SLOT(changeGUI(QString)));
#endif

#if SINGLE_INSTANCE
	MyApplication * app = MyApplication::instance();
	connect(app, SIGNAL(messageReceived(const QString&)),
            gui, SLOT(handleMessageFromOtherInstances(const QString&)));
	app->setActivationWindow(gui);
#endif

	return gui;
}

void SMPlayer::deleteGUI() {
#ifdef LOG_SMPLAYER
	allow_to_send_log_to_gui = false;
#endif

	delete main_window;
	main_window = 0;

#ifdef LOG_SMPLAYER
	allow_to_send_log_to_gui = true;
#endif
}

#ifdef GUI_CHANGE_ON_RUNTIME
void SMPlayer::changeGUI(QString new_gui) {
	qDebug("SMPlayer::changeGUI: '%s'", new_gui.toLatin1().constData());

	deleteGUI();

	main_window = createGUI(new_gui);

	main_window->show();
}
#endif

SMPlayer::ExitCode SMPlayer::processArgs(QStringList args) {
	qDebug("SMPlayer::processArgs: arguments: %d", args.count());
	for (int n = 0; n < args.count(); n++) {
		qDebug("SMPlayer::processArgs: %d = %s", n, args[n].toUtf8().data());
	}


	QString action; // Action to be passed to running instance
	bool show_help = false;

	if (!pref->gui.isEmpty()) gui_to_use = pref->gui;
	bool add_to_playlist = false;

#ifdef Q_OS_WIN
	if (args.contains("-uninstall")) {
		#if USE_ASSOCIATIONS
		//Called by uninstaller. Will restore old associations.
		WinFileAssoc RegAssoc; 
		Extensions exts; 
		QStringList regExts; 
		RegAssoc.GetRegisteredExtensions(exts.multimedia(), regExts); 
		RegAssoc.RestoreFileAssociations(regExts); 
		printf("Restored associations\n");
		#endif
		return NoError;
	}
#endif

	if (args.contains("-delete-config")) {
		CleanConfig::clean(Paths::configPath());
		return NoError;
	}

	for (int n = 1; n < args.count(); n++) {
		QString argument = args[n];

		if (argument == "-send-action") {
			if (n+1 < args.count()) {
				n++;
				action = args[n];
			} else {
				printf("Error: expected parameter for -send-action\r\n");
				return ErrorArgument;
			}
		}
		else
		if (argument == "-actions") {
			if (n+1 < args.count()) {
				n++;
				actions_list = args[n];
			} else {
				printf("Error: expected parameter for -actions\r\n");
				return ErrorArgument;
			}
		}
		else
		if (argument == "-sub") {
			if (n+1 < args.count()) {
				n++;
				QString file = args[n];
				if (QFile::exists(file)) {
					subtitle_file = QFileInfo(file).absoluteFilePath();
				} else {
					printf("Error: file '%s' doesn't exists\r\n", file.toUtf8().constData());
				}
			} else {
				printf("Error: expected parameter for -sub\r\n");
				return ErrorArgument;
			}
		}
		else
		if (argument == "-start") {
			if (n+1 < args.count()) {
				n++;
				QString str_pos = args[n];
				int h = 0;
				int m = 0;
				int s = 0;
				QRegExp rx("(\\d+):(\\d+):(\\d+)");
				if (rx.indexIn(str_pos) != -1) {
					h = rx.cap(1).toInt();
					m = rx.cap(2).toInt();
					s = rx.cap(3).toInt();
				} else {
					rx.setPattern("(\\d+):(\\d+)");
					if (rx.indexIn(str_pos) != -1) {
						m = rx.cap(1).toInt();
						s = rx.cap(2).toInt();
					}
					else s = str_pos.toInt();
				}
				//qDebug("SMPlayer::processArgs: %d:%d:%d", h, m, s);
				initial_second = (h * 60 * 60) + (m * 60) + s;
				qDebug("SMPlayer::processArgs: initial_second: %d", initial_second);
			} else {
				printf("Error: expected parameter for -start\r\n");
				return ErrorArgument;
			}
		}
		else
		if (argument == "-media-title") {
			if (n+1 < args.count()) {
				n++;
				if (media_title.isEmpty()) media_title = args[n];
			}
		}
		else
		if (argument == "-pos") {
			if (n+2 < args.count()) {
				bool ok_x, ok_y;
				n++;
				gui_position.setX( args[n].toInt(&ok_x) );
				n++;
				gui_position.setY( args[n].toInt(&ok_y) );
				if (ok_x && ok_y) move_gui = true;
			} else {
				printf("Error: expected parameter for -pos\r\n");
				return ErrorArgument;
			}
		}
		else
		if (argument == "-size") {
			if (n+2 < args.count()) {
				bool ok_width, ok_height;
				n++;
				gui_size.setWidth( args[n].toInt(&ok_width) );
				n++;
				gui_size.setHeight( args[n].toInt(&ok_height) );
				if (ok_width && ok_height) resize_gui = true;
			} else {
				printf("Error: expected parameter for -resize\r\n");
				return ErrorArgument;
			}
		}
		else
		if ((argument == "--help") || (argument == "-help") ||
            (argument == "-h") || (argument == "-?") ) 
		{
			show_help = true;
		}
		else
		if (argument == "-close-at-end") {
			close_at_end = 1;
		}
		else
		if (argument == "-no-close-at-end") {
			close_at_end = 0;
		}
		else
		if (argument == "-fullscreen") {
			start_in_fullscreen = 1;
		}
		else
		if (argument == "-no-fullscreen") {
			start_in_fullscreen = 0;
		}
		else
		if (argument == "-add-to-playlist") {
			add_to_playlist = true;
		}
		else
		if (argument == "-mini" || argument == "-minigui") {
			gui_to_use = "MiniGUI";
		}
		else
		if (argument == "-mpcgui") {
			gui_to_use = "MpcGUI";
		}
		else
		if (argument == "-defaultgui") {
			gui_to_use = "DefaultGUI";
		}
		else
		if (argument == "-ontop") {
			pref->stay_on_top = Preferences::AlwaysOnTop;
		}
		else
		if (argument == "-no-ontop") {
			pref->stay_on_top = Preferences::NeverOnTop;
		}
#ifdef SKINS
		else
		if (argument == "-skingui") {
			gui_to_use = "SkinGUI";
		}
#endif
		else {
			// File
			#if QT_VERSION >= 0x040600
			QUrl fUrl = QUrl::fromUserInput(argument);
			if (fUrl.isValid() && fUrl.scheme().toLower() == "file") {
				argument = fUrl.toLocalFile();
			}
			#endif
			if (QFile::exists( argument )) {
				argument = QFileInfo(argument).absoluteFilePath();
			}
			files_to_play.append( argument );
		}
	}

	if (show_help) {
		printf("%s\n", CLHelp::help().toLocal8Bit().data());
		return NoError;
	}

	qDebug("SMPlayer::processArgs: files_to_play: count: %d", files_to_play.count() );
	for (int n=0; n < files_to_play.count(); n++) {
		qDebug("SMPlayer::processArgs: files_to_play[%d]: '%s'", n, files_to_play[n].toUtf8().data());
	}

#ifdef SINGLE_INSTANCE
	if (pref->use_single_instance) {
		// Single instance
		MyApplication * a = MyApplication::instance();
		if (a->isRunning()) {
			a->sendMessage("Hello");

			if (!action.isEmpty()) {
				a->sendMessage("action " + action);
			}
			else {
				if (!subtitle_file.isEmpty()) {
					a->sendMessage("load_sub " + subtitle_file);
				}

				if (initial_second != 0) {
					a->sendMessage("start_second " + QString::number(initial_second));
				}

				if (!media_title.isEmpty()) {
					a->sendMessage("media_title " + files_to_play[0] + " <<sep>> " + media_title);
				}

				if (!files_to_play.isEmpty()) {
					/* a->sendMessage("open_file " + files_to_play[0]); */
					QString command = "open_files";
					if (add_to_playlist) command = "add_to_playlist";
					a->sendMessage(command +" "+ files_to_play.join(" <<sep>> "));
				}
			}

			return NoError;
		}
	}
#endif

	/*
	if (!pref->default_font.isEmpty()) {
		QFont f;
		f.fromString(pref->default_font);
		qApp->setFont(f);
	}
	*/

	return SMPlayer::NoExit;
}

void SMPlayer::start() {
#ifdef FONTCACHE_DIALOG
#ifndef PORTABLE_APP
	if (Version::with_revision() != pref->smplayer_version) {
		FontCacheDialog d(0);
		d.run(pref->mplayer_bin, "sample.avi");
		pref->smplayer_version = Version::with_revision();
	}
#endif
#endif

	if (!gui()->startHidden() || !files_to_play.isEmpty() ) gui()->show();
	if (!files_to_play.isEmpty()) {
		if (!subtitle_file.isEmpty()) gui()->setInitialSubtitle(subtitle_file);
		if (!media_title.isEmpty()) gui()->getCore()->addForcedTitle(files_to_play[0], media_title);
		gui()->setInitialSecond(initial_second);
		gui()->openFiles(files_to_play);
	}

	if (!actions_list.isEmpty()) {
		if (files_to_play.isEmpty()) {
			gui()->runActions(actions_list);
		} else {
			gui()->runActionsLater(actions_list);
		}
	}
}

#ifndef PORTABLE_APP
void SMPlayer::createConfigDirectory() {
	// Create smplayer config directory
	if (!QFile::exists(Paths::configPath())) {
		QDir d;
		if (!d.mkdir(Paths::configPath())) {
			qWarning("SMPlayer::createConfigDirectory: can't create %s", Paths::configPath().toUtf8().data());
		}
		// Screenshot folder already created in preferences.cpp if Qt >= 4.4
		#if QT_VERSION < 0x040400 
		QString s = Paths::configPath() + "/screenshots";
		if (!d.mkdir(s)) {
			qWarning("SMPlayer::createHomeDirectory: can't create %s", s.toUtf8().data());
		}
		#endif
	}
}
#endif

#ifdef FONTS_HACK
void SMPlayer::createFontFile() {
	qDebug("SMPlayer::createFontFile");
	QString output = Paths::configPath() + "/fonts.conf";

	// Check if the file already exists with the modified path
	if (QFile::exists(output)) {
		QFile i(output);
		if (i.open(QIODevice::ReadOnly | QIODevice::Text)) {
			QString text = i.readAll();
			if (text.contains("<dir>" + Paths::fontPath() + "</dir>")) {
				qDebug("SMPlayer::createFontFile: file %s already exists with font path. Doing nothing.", output.toUtf8().constData());
				return;
			}
		}
	}

	QString input = Paths::appPath() + "/mplayer/fonts/fonts.conf";
	if (!QFile::exists(input)) {
		qDebug("SMPlayer::createFontFile: %s doesn't exist", input.toUtf8().constData());
		input = Paths::appPath() + "/mplayer/mpv/fonts.conf";
		if (!QFile::exists(input)) {
			qDebug("SMPlayer::createFontFile: %s doesn't exist", input.toUtf8().constData());
			qWarning("SMPlayer::createFontFile: failed to create fonts.conf");
			return;
		}
	}
	qDebug("SMPlayer::createFontFile: input: %s", input.toUtf8().constData());
	QFile infile(input);
	if (infile.open(QIODevice::ReadOnly | QIODevice::Text)) {
		QString text = infile.readAll();
		text = text.replace("<!-- <dir>WINDOWSFONTDIR</dir> -->", "<dir>WINDOWSFONTDIR</dir>");
		text = text.replace("<dir>WINDOWSFONTDIR</dir>", "<dir>" + Paths::fontPath() + "</dir>");
		//qDebug("SMPlayer::createFontFile: %s", text.toUtf8().constData());

		qDebug("SMPlayer::createFontFile: saving %s", output.toUtf8().constData());
		QFile outfile(output);
		if (outfile.open(QIODevice::WriteOnly | QIODevice::Text)) {
			outfile.write(text.toUtf8());
			outfile.close();
		}
	}
}
#endif

void SMPlayer::showInfo() {
#ifdef Q_OS_WIN
	QString win_ver;
	switch (QSysInfo::WindowsVersion) {
		case QSysInfo::WV_2000: win_ver = "Windows 2000"; break;
		case QSysInfo::WV_XP: win_ver = "Windows XP"; break;
		case QSysInfo::WV_2003: win_ver = "Windows XP Professional x64/Server 2003"; break;
		case QSysInfo::WV_VISTA: win_ver = "Windows Vista/Server 2008"; break;
		#if QT_VERSION >= 0x040501
		case QSysInfo::WV_WINDOWS7: win_ver = "Windows 7/Server 2008 R2"; break;
		#endif
		#if QT_VERSION >= 0x040803
		case QSysInfo::WV_WINDOWS8: win_ver = "Windows 8/Server 2012"; break;
		#endif
		#if ((QT_VERSION >= 0x040806 && QT_VERSION < 0x050000) || (QT_VERSION >= 0x050200))
		case QSysInfo::WV_WINDOWS8_1: win_ver = "Windows 8.1/Server 2012 R2"; break;
		#endif
		#if ((QT_VERSION >= 0x040807 && QT_VERSION < 0x050000) || (QT_VERSION >= 0x050500))
		case QSysInfo::WV_WINDOWS10: win_ver = "Windows 10"; break;
		#endif
		case QSysInfo::WV_NT_based: win_ver = "NT-based Windows"; break;
		default: win_ver = QString("Unknown/Unsupported Windows OS"); break;
	}
#endif
	QString s = QObject::tr("This is SMPlayer v. %1 running on %2")
            .arg(Version::printable())
#ifdef Q_OS_LINUX
           .arg("Linux")
#else
#ifdef Q_OS_WIN
           .arg("Windows ("+win_ver+")")
#else
#ifdef Q_OS_OS2
           .arg("eCS (OS/2)")
#else
		   .arg("Other OS")
#endif
#endif
#endif
           ;

	printf("%s\n", s.toLocal8Bit().data() );
	qDebug("%s", s.toUtf8().data() );
	qDebug("Compiled with Qt v. %s, using %s", QT_VERSION_STR, qVersion());

	qDebug() << " * application path:" << Paths::appPath();
	qDebug() << " * data path:" << Paths::dataPath();
	qDebug() << " * translation path:" << Paths::translationPath();
	qDebug() << " * doc path:" << Paths::docPath();
	qDebug() << " * themes path:" << Paths::themesPath();
	qDebug() << " * shortcuts path:" << Paths::shortcutsPath();
	qDebug() << " * config path:" << Paths::configPath();
	qDebug() << " * ini path:" << Paths::iniPath();
	qDebug() << " * file for subtitles' styles:" << Paths::subtitleStyleFile();
	qDebug() << " * current path:" << QDir::currentPath();
#ifdef FONTS_HACK
	qDebug() << " * font path:" << Paths::fontPath();
#endif
}

#ifdef LOG_SMPLAYER
QFile SMPlayer::output_log;
bool SMPlayer::allow_to_send_log_to_gui = false;

#if QT_VERSION >= 0x050000
void SMPlayer::myMessageOutput( QtMsgType type, const QMessageLogContext &, const QString & msg ) {
#else
void SMPlayer::myMessageOutput( QtMsgType type, const char *msg ) {
#endif
	static QStringList saved_lines;
	static QString orig_line;
	static QString line2;
	static QRegExp rx_log;

	if (pref) {
		if (!pref->log_smplayer) return;
		rx_log.setPattern(pref->log_filter);
	} else {
		rx_log.setPattern(".*");
	}

	line2.clear();

#if QT_VERSION >= 0x050000
	orig_line = msg;
#else
	#ifdef Q_OS_WIN
	orig_line = QString::fromLocal8Bit(msg);
	#else
	orig_line = QString::fromUtf8(msg);
	#endif
#endif

	switch ( type ) {
		case QtDebugMsg:
			if (rx_log.indexIn(orig_line) > -1) {
				#ifndef NO_DEBUG_ON_CONSOLE
				fprintf( stderr, "Debug: %s\n", orig_line.toLocal8Bit().data() );
				#endif
				line2 = orig_line;
			}
			break;
		case QtWarningMsg:
			#ifndef NO_DEBUG_ON_CONSOLE
			fprintf( stderr, "Warning: %s\n", orig_line.toLocal8Bit().data() );
			#endif
			line2 = "WARNING: " + orig_line;
			break;
		case QtFatalMsg:
			#ifndef NO_DEBUG_ON_CONSOLE
			fprintf( stderr, "Fatal: %s\n", orig_line.toLocal8Bit().data() );
			#endif
			line2 = "FATAL: " + orig_line;
			abort();                    // deliberately core dump
		case QtCriticalMsg:
			#ifndef NO_DEBUG_ON_CONSOLE
			fprintf( stderr, "Critical: %s\n", orig_line.toLocal8Bit().data() );
			#endif
			line2 = "CRITICAL: " + orig_line;
			break;
		#if QT_VERSION >= 0x050500
		case QtInfoMsg:
			break;
		#endif
	}

	if (line2.isEmpty()) return;

	line2 = "["+ QTime::currentTime().toString("hh:mm:ss:zzz") +"] "+ line2;

	if (allow_to_send_log_to_gui && main_window) {
		if (!saved_lines.isEmpty()) {
			// Send saved lines first
			for (int n=0; n < saved_lines.count(); n++) {
				main_window->recordSmplayerLog(saved_lines[n]);
			}
			saved_lines.clear();
		}
		main_window->recordSmplayerLog(line2);
	} else {
		// GUI is not created yet, save lines for later
		saved_lines.append(line2);
		/* printf("SMPlayer::myMessageOutput: no gui\n"); */
	}

	if (pref) {
		if (pref->save_smplayer_log) {
			// Save log to file
			if (!output_log.isOpen()) {
				// FIXME: the config path may not be initialized if USE_LOCKS is not defined
				output_log.setFileName( Paths::configPath() + "/smplayer_log.txt" );
				output_log.open(QIODevice::WriteOnly);
			}
			if (output_log.isOpen()) {
				QString l = line2 + "\r\n";
				output_log.write(l.toUtf8().constData());
				output_log.flush();
			}
		}
	}
}
#endif

/*
void myMessageOutput( QtMsgType type, const char *msg ) {
	static QString orig_line;
	orig_line = QString::fromUtf8(msg);

	switch ( type ) {
		case QtDebugMsg:
			#ifndef NO_DEBUG_ON_CONSOLE
			fprintf( stderr, "Debug: %s\n", orig_line.toLocal8Bit().data() );
			#endif
			break;

		case QtWarningMsg:
			#ifndef NO_DEBUG_ON_CONSOLE
			fprintf( stderr, "Warning: %s\n", orig_line.toLocal8Bit().data() );
			#endif
			break;

		case QtCriticalMsg:
			#ifndef NO_DEBUG_ON_CONSOLE
			fprintf( stderr, "Critical: %s\n", orig_line.toLocal8Bit().data() );
			#endif
			break;

		case QtFatalMsg:
			#ifndef NO_DEBUG_ON_CONSOLE
			fprintf( stderr, "Fatal: %s\n", orig_line.toLocal8Bit().data() );
			#endif
			abort();                    // deliberately core dump
	}
}
*/

#include "moc_smplayer.cpp"
